New VRF Heat Pump Model

rongpengzhang opened the pull request:
##### Introduction of the New VRF-HP Model

The proposed new feature introduces a new model to simulate the energy performance of Variable Refrigerant Flow (VRF) systems in the heat pump (HP) operation mode (either cooling or heating is provided but not simultaneously). 

Compared with the existing VRF-HP model in EnergyPlus, the proposed model is more physics-based and thus able to consider the dynamics of more operational parameters, which is essential for the description of many enhanced control logics (e.g., the adjustment of superheating degrees during small load operations). Furthermore, the proposed model implements component-level curves rather than the system-level curves, and thus requires fewer curves as model inputs. The definition of VRF performance data for simulation as part of ASHRAE Standard 205 will ensure VRF manufacturers provide adequate data to create the required performance curves.

The new VRF model will be an alternative for the users. We propose to add the new VRF heat pump model in EnergyPlus for the Fall 2015 release, while keeping the existing VRF model intact. An on-going effort covers the development and validation of a new VRF heat recovery (HR) model, which will be implemented in future release (2016) of EnergyPlus.
##### Code Hierarchy for the New VRF-HP Model

![vrf-hp coding hierarhcy_zrp](https://cloud.githubusercontent.com/assets/7785946/9286859/dbf7ef74-42b2-11e5-8c3c-9965ca86fcfb.PNG)

rongpengzhang committed:
VRF-HP: V1.1+V1.2

V1.1: Add new IDD objects: (fields old) eg. •
Coil:Cooling:DX:VariableRefrigerantFlow:FluidTemperatureControl •
AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl
V1.2: Src code:  new method for the simulation using new IDD objects:
eg. CalcVRF_FluidTCtrl

rongpengzhang committed:
VRF-HP: V1.3.1+V1.3.2

V1.3 Src code:  VRF data struct, new member to distinguish old and new
algorithm.  eg. CalcVRF_FluidTCtrl

rongpengzhang committed:
VRF-HP: V1.3.3

V1.3 Src code:  VRF data struct, new member to distinguish old and new
algorithm.  eg. CalcVRF_FluidTCtrl

rongpengzhang committed:
VEF-HP: V1.4 + V2.2.a

V1.4: New methods for DX Coil
V2.2.a Move hh file part for VRF: member definition

rongpengzhang committed:
VRF-HP: V1.5

V1.5 Data Struct Update, method connections, IDF Read in for DX Coil

rongpengzhang committed:
VRF-HP: V2.1.a

V2.1.a Add New methods C3, C2, C2.1, C2.2
C3 CalcVRFEvapCondTemp
C2 CalcVRFAirFlow
C2.1 FanSpdResidual
C2.2 FanSpdResidualHeat

rongpengzhang committed:
VRF-HP: V2.1.b

V2.1.b Data Struct Update

rongpengzhang committed:
VRF-HP: V2.1.c

V2.1.c Finish method C1, C1.2(d)
C1 SimDXCoil
C1.2 CalcVRFHeatingCoil

rongpengzhang committed:
VRF-HP: V2.1.d

V2.1 d C1.1(c)
C1.1 CalcVRFCoolingCoil_FluidTCtrl method is added

rongpengzhang committed:
VRF-HP: V2.1.e

Update method C1.3 GetDXCoils

rongpengzhang committed:
VRF-HP: V2.2.b

modify A SimulateVRF
add A9 CompResidual

rongpengzhang committed:
VRF-HP: V2.2.c-1

V2.2.c  Finish A5 CalcVRFCondenser (Cooling Mode)

rongpengzhang committed:
VRF-HP: V2.2.c-2

Finish A5: method CalcVRFCondenser_FluidTCtrl
Finish A2.2: method CalcVRF_FluidTCtrl

rongpengzhang committed:
VRF-HP: V2.2.d

V2.2.d  Others: A1, A4, A2, A2.1
A1: GetVRFInput
A4: ReportVRFTerminalUnit
A2: InitVRF

Myoldmopar commented:
@rraustad Are you able to be the reviewer/merger on this one?  I can do it, but given your _familiarity_ with the existing VRF model, I'd appreciate your explicit approval.

rongpengzhang commented:
@Myoldmopar @rraustad This feature is under development. I will notify you when it is done, and I will provide some code hierarchy charts which can easy the review.

rraustad commented:
I'll try to be ready to review when this is finalized.

rongpengzhang committed:
VRF-HP: V3.1.a

IDD update. Create Unit tests. Some other minor changes.

rongpengzhang committed:
VRF-HP: V3.1.b-

Files changed to previous version to avoid merge conflicts

rongpengzhang committed:
VRF-HP: Deal with merge conflicts

rongpengzhang committed:
Merge remote-tracking branch 'origin/develop' into 91977808_LBNL_NewVRFModel

rongpengzhang committed:
VRF-HP: V3.1.b

New parameters linking DXCoil and VRFIU/OU
Merger from the develop branch

rongpengzhang committed:
VRF-HP: V3.2.a

Unit test developed and Codes debugged for Cooling Mode

rongpengzhang committed:
VRF-HP: V3.2.b

Codes clean up and bug fixed.

rongpengzhang committed:
VRF-HP: V3.2.c

Create new method CalcVRFIUTeTc_FluidTCtrl

rongpengzhang committed:
Merge remote-tracking branch 'origin/develop' into 91977808_LBNL_NewVRFModel

rongpengzhang committed:
VRF-HP: V3.2.c +

Deal with all the GitHub Build Warnings

rongpengzhang committed:
VRF-HP: V4.1

Implement the IDD proposed, including
Coil:Cooling:DX:VariableRefrigerantFlow:FluidTemperatureControl,
Coil:Heating:DX:VariableRefrigerantFlow:FluidTemperatureControl, and
AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl.

rongpengzhang committed:
VRF-HP: V4.2

Deal with warnings. Most are from the refrigerant property calculations.
Value may be out of range during iterations.

rraustad commented:
The branch has conflicts which should be merged with develop. Please perform the merge while I review these changes.

There is 1 failed unit test, and several files with diff's. I cannot see what those diffs are so I cannot verify something isn't wrong without running those files locally. Did you run the test suite locally and have the csv results before and after your changes? If so, could you send my the VariableRefrigerantFlow_5Zone csv's? It's OK for the answer to change as long as the fundamental operation of the system hasn't changed. For example, zone temp's are still maintained the same, coil heating/cooling rate vs time hasn't changed, air flow rate vs time hasn't changed, etc.

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -4866,6 +4885,185 @@ namespace DXCoils {
+			// @@ Use fan object data
+			DXCoil( DXCoilNum ).RatedAirVolFlowRate( 1 ) = 0.1793;
+			DXCoil( DXCoilNum ).Qfan = 35.1;
+			DXCoil( DXCoilNum ).BF = 0.0592;
```
Why are these variables hard-coded?

rongpengzhang replied:
They were hard-coded temporarily. Now these values are obtained from fan object calculations.

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -4866,6 +4885,185 @@ namespace DXCoils {
+			// @@ Use fan object data
+			DXCoil( DXCoilNum ).RatedAirVolFlowRate( 1 ) = 0.1793;
+			DXCoil( DXCoilNum ).Qfan = 35.1;
+			DXCoil( DXCoilNum ).BF = 0.136;
```
Why are these values hard-coded?

rongpengzhang replied:
They were hard-coded temporarily. Now these values are obtained from fan object calculations.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -33425,6 +33425,14 @@ ZoneHVAC:TerminalUnit:VariableRefrigerantFlow,
+        \note Supply Air Fan Object Type must be Fan:VariableVolume if
+        \note AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl
+        \note is used to model VRF outdoor unit
         \default Fan:ConstantVolume
```
This should be tested in GetInput so that users cannot use the wrong fan type.

rongpengzhang replied:
Tests are added in GetVRFInput().

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -33441,7 +33449,14 @@ ZoneHVAC:TerminalUnit:VariableRefrigerantFlow,
+        \note Cooling Coil Type must be 
+        \note Coil:Cooling:DX:VariableRefrigerantFlow:FluidTemperatureControl
+        \note if AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl 
+        \note is used to model VRF outdoor unit
```
This should be tested in GetInput so that users cannot use the wrong fan type.

rongpengzhang replied:
Tests are added in GetVRFInput().

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -33451,7 +33466,14 @@ ZoneHVAC:TerminalUnit:VariableRefrigerantFlow,
+        \note is used to model VRF outdoor unit
         \note This field may be left blank if cooling-only mode is used
   A14,  \field Heating Coil Object Name
         \type object-list
```
You need to add the \key HeatingCoilsDXVarRefrigFlowFluidTemperatureControl here

rongpengzhang replied:
\object-list HeatingCoilsDXVarRefrigFlowFluidTemperatureControl is added in IDD.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -33441,7 +33449,14 @@ ZoneHVAC:TerminalUnit:VariableRefrigerantFlow,
+        \note is used to model VRF outdoor unit
         \note This field may be left blank if heating-only mode is used
   A12,  \field Cooling Coil Object Name
         \type object-list
```
You need to add the \key CoolingCoilsDXVarRefrigFlowFluidTemperatureControl here

rongpengzhang replied:
\object-list CoolingCoilsDXVarRefrigFlowFluidTemperatureControl is added in IDD.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -42751,6 +42773,93 @@ Coil:Heating:DX:VariableRefrigerantFlow,
+  N3,  \field Indoor Unit Reference Superheating 
+       \required-field
+       \type real
+       \units deltaC 
```
Are any of these fields, N1-N3, autosizable? If not, it will be difficult to run simulations over different geographies.

rraustad replied:
Have you tried using autosize for Rated Total Cooling Capacity and Rated Sensible Heat Ratio? Is it possible to autosize these inputs? Can you add a default for Indoor Unit Reference Superheating since this value is typically around 5C? If you do add a default, remove the \required-field tag.

rongpengzhang replied:
I have updated the codes to activate the autosizing functions for N1 and N2. A default value is used for N3 and the \required-field tag is removed.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -42751,6 +42773,93 @@ Coil:Heating:DX:VariableRefrigerantFlow,
+       \required-field
+       \type real 
+       \units deltaC
+  A5;  \field Indoor Unit Condensing Temperature Function of Subcooling Curve Name    
```
Are any of these fields, N1-N2, autosizable? If not, it will be difficult to run simulations over different geographies.

rraustad replied:
Have you tried using autosize for Rated Total Heating Capacity? Is it possible to autosize this input? Can you add a default for Indoor Unit Reference Subcooling since this value is typically around 5C (check this)? If you do add a default, remove the \required-field tag.

rongpengzhang replied:
I have updated the codes to activate the autosizing functions for N1. A default value is used for N2 and the \required-field tag is removed.

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -14005,6 +14279,1379 @@ Label50: ;
+			// For multimode coil, this should be full flow including bypassed fraction
+			// AirMassFlow = DXCoil( DXCoilNum ).InletAirMassFlowRate;
+			// DXCoil( DXCoilNum ).TotalCoolingEnergyRate = AirMassFlow * ( InletAirEnthalpy - OutletAirEnthalpy );
+			
```
These commented out lines should be removed.

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -14005,6 +14279,1379 @@ Label50: ;
+			AirMassFlowRatio = AirMassFlow / DXCoil( DXCoilNum ).RatedAirMassFlowRate( Mode );
+			//TotCapFlowModFac = CurveValue( DXCoil( DXCoilNum ).CCapFFlow( Mode ),AirMassFlowRatio )
+			//@@
+			TotCapModFac = 1.0;
```
You should clean up all @@ reminders and remove the lines or charachters

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -14005,6 +14279,1379 @@ Label50: ;
+        Real64 SHRini;
+        Real64 DeltaT;
+        Real64 RHsat;
+		
```
No one will know what these variables represent nor their units. You really should add comments. I can live with some missing, but all are missing here.

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -14005,6 +14279,1379 @@ Label50: ;
+					SHact = 999.0;
+				}
+			}
+        } else {
```
These lines are not justified correctly. Indented too far.

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -14005,6 +14279,1379 @@ Label50: ;
+		//NO HEATING/COOLING: Mode 2
+        
+            TairInlet = DXCoil( CoolCoil ).InletAirTemp;
+            Wi
```
These lines are not justified correctly. Indented too far.

rraustad commented on tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc:
```diff
@@ -0,0 +1,361 @@
+
+	// Run and Check
+    double FanSpdResidual = FanSpdResidualHeat( FanSpdRto, Par );
+	EXPECT_NEAR( -0.5459, FanSpdResidual, 0.0005 );
```
These lines are not justified.

rongpengzhang replied:
Cleaned. This is caused by the use of Tab characters.

rongpengzhang commented:
Hi @rraustad, thanks for the review. In fact, I'm still in the final cleaning up step. It can be done in 1-2 days. I planned to notify and invite you to review after that. Can you please wait for a bit? I will upload the newest version soon.

rongpengzhang committed:
VRF-HP: V4.2 a-f

Clean up.

rongpengzhang committed:
Merge branch 'develop' into 91977808_LBNL_NewVRFModel

Conflicts:
	src/EnergyPlus/DXCoils.hh

rraustad commented:
OK, that's fine. I just wanted to make a quick pass so you would have 
time to make changes.

On 8/12/2015 5:06 PM, Rongpeng Zhang wrote:

> Hi @rraustad https://github.com/rraustad, thanks for the review. In 
> fact, I'm still in the final cleaning up step. It can be done in 1-2 
> days. I planned to notify and invite you to review after that. Can you 
> please wait for a bit? I will update the newest version soon.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/NREL/EnergyPlus/pull/5026#issuecomment-130449168.

## 

Richard Raustad
Florida Solar Energy Center
1679 Clearlake Road
Cocoa, FL 32922
Ph: (321)638-1454
http://www.fsec.ucf.edu/en/

Program Director
Electric Vehicle Transportation Center
http://evtc.fsec.ucf.edu/

rongpengzhang committed:
VRF-HP: V4.2 f-g

Clean up DXCoil and VRF.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52927,6 +53038,337 @@ AirConditioner:VariableRefrigerantFlow,
+       \required-field
+       \type object-list
+       \object-list FluidNames
+       \default R410A
```
If there is a default value this input IS NOT a \required-field. Remove the \required-field tag.

rongpengzhang replied:
The \required-field tag is removed.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52927,6 +53038,337 @@ AirConditioner:VariableRefrigerantFlow,
+  N4 , \field Reference Outdoor Unit Subcooling 
+       \required-field
+       \type real 
+       \units deltaC   
```
Can you add a default for N3 and N4? If so, remove the \required-field tags.

rongpengzhang replied:
Default N3 and N4 values are provided.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52927,6 +53038,337 @@ AirConditioner:VariableRefrigerantFlow,
+       \type choice
+       \key ConstantTemp
+       \key VariableTemp
+       \default VariableTemp
```
Remove the \required-field tag.

rongpengzhang replied:
The tag is removed.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52927,6 +53038,337 @@ AirConditioner:VariableRefrigerantFlow,
+       \type object-list
+       \object-list QuadraticCurves	 
+       \object-list UniVariateTables  
+  N11, \field Diameter of main pipe connecting outdoor unit to indoor units
```
Just checking here that these 2 curves are required and cannot be left blank.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52927,6 +53038,337 @@ AirConditioner:VariableRefrigerantFlow,
+       \required-field   
+       \type object-list
+       \object-list QuadraticCurves	 
+       \object-list UniVariateTables  
```
Just checking here that these 2 curves are required and cannot be left blank.

rongpengzhang replied:
Yes, these two curves are required from the user.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52927,6 +53038,337 @@ AirConditioner:VariableRefrigerantFlow,
+       \note Positive means outdoor unit is higher than indoor unit 
+       \note Negative means outdoor unit is lower than indoor unit 
+       \type real 
+       \units m
```
\minimum 0 here?

rongpengzhang replied:
This field can be either positive or negative. Positive means outdoor unit is higher than indoor unit , while negative means outdoor unit is lower than indoor unit.

rraustad replied:
Yes, you are correct. I was focusing more on idd syntax than the field itself.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52927,6 +53038,337 @@ AirConditioner:VariableRefrigerantFlow,
+       \required-field
+       \type real 
+       \units W/m-K
+       \minimum 0.0	
```
It would be better if field N11 - N16 were not required fields and all had a default. I won't force this. For example a 1" pipe diameter and a 30 m pipe length default. Whatever you think for the other inputs. These could be described in the IO Reference for each field.

rongpengzhang replied:
It is a good idea to use default values. Thanks for the suggestion!

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52927,6 +53038,337 @@ AirConditioner:VariableRefrigerantFlow,
+       \units kPa
+       \default 4500
+       \minimum 0.0
+       \maximum 50000
```
Remove \required-field tag.

rongpengzhang replied:
Tag removed.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52927,6 +53038,337 @@ AirConditioner:VariableRefrigerantFlow,
+       \minimum 2
+       \maximum 9
+       \note First index represents minimal capacity operation 
+       \note Last index represents full capacity operation 
```
Remove \required-field tag.

rongpengzhang replied:
Tag removed.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52927,6 +53038,337 @@ AirConditioner:VariableRefrigerantFlow,
+       \required-field
+       \type object-list
+       \object-list BiQuadraticCurves
+       \object-list BiVariateTables
```
Verify A13 and A14 are required fields.

rongpengzhang replied:
Yes, these fields are required. Codes were designed in GetVRFInput() to ensure that.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52927,6 +53038,337 @@ AirConditioner:VariableRefrigerantFlow,
+       \required-field
+       \type object-list
+       \object-list BiQuadraticCurves
+       \object-list BiVariateTables
```
Verify A15 and A16 are required fields.

rongpengzhang replied:
Yes, these fields are required. Codes were designed in GetVRFInput() to ensure that.

rraustad commented on src/EnergyPlus/DataHVACGlobals.cc:
```diff
@@ -166,6 +165,10 @@ namespace DataHVACGlobals {
+	int const CoilDX_HeatPumpWaterHeaterWrapped( 31 );
+	
+	int const CoilVRF_FluidTCtrl_Cooling( 32 );
+	int const CoilVRF_FluidTCtrl_Heating( 33 );
```
I reordered all these to be 1 - n, so we will have merge conflicts here (you, I, and Bereket). Just be careful if you have to merge DataHVACGlobals since these numbers are now in order.

rongpengzhang replied:
I successfully merged the develop branch into mine several days ago. In the merge before that, about 15 days ago, there were merge conflicts and I modified my codes to avoid the conflicts. Anyway, I will be very careful when merge.

rraustad replied:
Nice job, merging is a pain!

rraustad commented on src/EnergyPlus/DataHVACGlobals.cc:
```diff
@@ -178,7 +181,7 @@ namespace DataHVACGlobals {
 	int const WaterConstantOnDemand( 3 ); // water flow is constant whenever the coil is operational - this is the only method used in EP V7.2 and earlier
 
-	Array1D_string const cAllCoilTypes( NumAllCoilTypes, { "Coil:Cooling:DX:SingleSpeed", "Coil:Heating:DX:SingleSpeed", "Coil:Cooling:DX:TwoSpeed", "CoilSystem:Cooling:DX:HeatExchangerAssisted", "Coil:Cooling:DX:TwoStageWithHumidityControlMode", "Coil:WaterHeating:AirToWaterHeatPump:Pumped", "Coil:Cooling:DX:MultiSpeed", "Coil:Heating:DX:MultiSpeed", "Coil:Heating:Gas", "Coil:Heating:Gas:MultiStage ...
+	Array1D_string const cAllCoilTypes( NumAllCoilTypes, { "Coil:Cooling:DX:SingleSpeed", "Coil:Heating:DX:SingleSpeed", "Coil:Cooling:DX:TwoSpeed", "CoilSystem:Cooling:DX:HeatExchangerAssisted", "Coil:Cooling:DX:TwoStageWithHumidityControlMode", "Coil:WaterHeating:AirToWaterHeatPump:Pumped", "Coil:Cooling:DX:MultiSpeed", "Coil:Heating:DX:MultiSpeed", "Coil:Heating:Gas", "Coil:Heating:Gas:MultiStage ...
```
I also reordered these coil types so they would be in the same order as the 1-33 above. be careful when merging.

rongpengzhang replied:
Yes, I will. Thanks for the kind reminder.

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.hh:
```diff
@@ -293,10 +296,54 @@ namespace HVACVariableRefrigerantFlow {
 		Real64 EMSValueForHPOperatingMode;
 		int HPOperatingModeErrorIndex;
+		
+		//The following are for the Algorithm Type: VRF model based on physics, appliable for Fluid Temperature Control
```
...,applicable is misspelled

rongpengzhang replied:
Typos corrected.

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.hh:
```diff
@@ -293,10 +296,54 @@ namespace HVACVariableRefrigerantFlow {
+		Real64 OUCondTempHigh; // VRV system outdoor unit condensing temperature, higher bound [C]
+		Real64 OUAirFlowRate; // Max condenser air flow rate  [m3/s]
+		Real64 SH; // VRF outdoor unit uperheating degrees [C]
+		Real64 SC; // VRF outdoor unit ubcooling
```
uperheating and ubcooling are both misspelled

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.hh:
```diff
@@ -293,10 +296,54 @@ namespace HVACVariableRefrigerantFlow {
+		Array1D< Real64 > CompressorSpeed; // compressor speed array [rps]
+		Array1D_int OUCoolingCAPFT; // index to outdoor unit cooling capacity funtion of temperature at different compressor speed
+		Array1D_int OUCoolingPWRFT; // index to outdoor unit cooling power funtion of temperature at different compressor speed
+		Real64 CompMaxDeltaP; // maximum compressor pressure rise [Pa]
```
function is misspelled in both lines above

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.hh:
```diff
@@ -293,10 +296,54 @@ namespace HVACVariableRefrigerantFlow {
+		Real64 NcompHeating; // compressor electric power at heating mode [W]
+		Array1D< Real64 > CompressorSpeed; // compressor speed array [rps]
+		Array1D_int OUCoolingCAPFT; // index to outdoor unit cooling capacity funtion of temperature at different compressor speed
+		Array1D_int OUCoolingPWRFT; // index to outdoor unit cooling power funtion of temperature at different compressor speed
```
function is misspelled in both lines above

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.hh:
```diff
@@ -454,13 +501,51 @@ namespace HVACVariableRefrigerantFlow {
 			EMSValueForHPOperatingMode( 0.0 ),
-			HPOperatingModeErrorIndex( 0 )
+			HPOperatingModeErrorIndex( 0 ),
+			CondensingTemp( 25.0 ),
```
why is this 25C and not 35C? Just checking on initialization.

rongpengzhang replied:
I change it to 42.0. Also checked the other initialization values.

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.hh:
```diff
@@ -454,13 +501,51 @@ namespace HVACVariableRefrigerantFlow {
+			CondensingTemp( 25.0 ),
+			EvaporatingTemp( 6.0 ),
+			IUEvaporatingTemp( 6.0 ), 
+			IUCondensingTemp( 25.0 ),
```
I thought this and the one above (line 505) defaulted to 44C?

rongpengzhang replied:
Yes, 44 is a more reasonable value.

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.hh:
```diff
@@ -1343,10 +1511,42 @@ namespace HVACVariableRefrigerantFlow {
+		Real64 const PartLoadRatio, // compressor part load fraction
+		Real64 & LoadMet, // load met by unit (W)
+		Real64 & OnOffAirFlowRatio, // ratio of ON air flow to average air flow
+		Optional< Real64 > LatOutputProvided = _ // delivered latent capacity (W)
```
FYI, we are trying to avoid optional arguments in the future. I try to avoid this, just try to minimize optional arguments in the future.

rongpengzhang replied:
Thanks. I will keep this in mind.

rraustad commented on tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc:
```diff
@@ -0,0 +1,354 @@
+    	
+    using namespace DXCoils;
+	using namespace DataZoneEnergyDemands;
+	using namespace EnergyPlus::Psychrometrics;
```
alignment. add 1 space before above 2 lines

rongpengzhang replied:
Solved.

rraustad commented on tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc:
```diff
@@ -0,0 +1,354 @@
+    DXCoil( CoolCoilIndex ).C2Te = 0.804;
+    DXCoil( CoolCoilIndex ).C3Te = 0;
+    DXCoil( CoolCoilIndex ).SH	 = 3.00;
+	DXCoil( CoolCoilIndex ).SupplyFanIndex = 0;
```
alignment

rraustad commented on tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc:
```diff
@@ -0,0 +1,354 @@
+    DXCoil( HeatCoilIndex ).C2Tc = 0.4333;
+    DXCoil( HeatCoilIndex ).C3Tc = 0.0207;
+    DXCoil( HeatCoilIndex ).SC	 = 5.00;
+	DXCoil( HeatCoilIndex ).SupplyFanIndex = 0;
```
alignment

rraustad commented on tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc:
```diff
@@ -0,0 +1,354 @@
+    TerminalUnitList.deallocate( ); 
+}
+
+}
```
Great Job!!!

rongpengzhang replied:
Hi @rraustad ,

Thanks for the careful review and insightful comments! Really appreciate that! I will update the codes to address your comments soon.

(1) The chart showing the code hierarchy can be found via
https://www.dropbox.com/s/5xd05e49rx6r03b/VRF-HP%20Coding%20Hierarhcy_zrp.PNG?dl=0
(2) I have checked the 4 files that created diffs. I run them with the develop branch codes and my codes, and found the simulation results (eso file) exactly the same. I think the difference comes from the new added output variables. I will have further look at this. Details can be found via 
https://www.dropbox.com/sh/lq4u60gtxlgta50/AADTqRoPX1P4Zpla_dTMZUOXa?dl=0

Best,
Rongpeng

@hongtz68

rongpengzhang replied:
Code Hierarchy for the New VRF-HP Model
![vrf-hp coding hierarhcy_zrp](https://cloud.githubusercontent.com/assets/7785946/9286855/a77133b4-42b2-11e5-880b-19f9f6fcadbe.PNG)

rongpengzhang committed:
VRF-HP: V4.2+

Deal with CI test warnings and diffs.

rongpengzhang committed:
VRF-HP: V4.2 H-K

Address Rich's comments; Activate the autosizing function for the new
coil types

rongpengzhang committed:
VRF-HP: V4.2 L

CoolingPLR and HeatingPLR, Defrost electricity.

rongpengzhang committed:
VRF-HP: V4.2 L+

CrankCaseHeaterPower, OperatingHeatingCOP & OperatingCoolingCOP

rongpengzhang committed:
VRF-HP: V4.2 M

rongpengzhang committed:
VRF-HP: V4.2N

Add the example file; Add MinOATemperature for VRF operations.

rongpengzhang committed:
Merge branch 'develop' into 91977808_LBNL_NewVRFModel

rongpengzhang commented:
Hi @rraustad ,

I have finished updating the codes to address your comments. The codes have passed the testes (as what we noticed, 4 files failed due to minor changes but simulation results are the same).

If you have no further comments, I think we can get this branch merged today or tomorrow (I will try to finish the documentation today).  I have fixed all the merge conflicts when I merge the develop branch into mine, so there should be no merge conflict when merge it into develop.

There is no field modifications in existing objects, so I don't think we need to update the transition for this feature.

FYI, @hongtz68  @Myoldmopar

rongpengzhang committed:
VRF-HP: IO Ref and Eng Ref

rongpengzhang committed:
Merge branch 'develop' into 91977808_LBNL_NewVRFModel

rongpengzhang committed:
VRF_HP: Unit Test

rraustad commented on doc/src/docs/EngineeringReference/13c-EncyclopaedicRefs.md:
```diff
@@ -5362,13 +5373,1052 @@ where:
+Note that a number of calculation steps are coupled together in the VRF_FluidTCtrl model, for instance, the piping loss calculation and the system performance calculation. More specifically, the piping loss changes the operating conditions of the system, which may lead to a different control strategy and thus in reverse affect the amount of piping loss. This makes it difficult to obtain an analyt ...
+
+The object connections for VRF_FluidTCtrl model is similar to those for VRF-SysCurve model. The difference lies in the object used to describe a specific components. For example, VRF-SysCurve model uses  *AirConditioner:VariableRefrigerantFlow* object to describe the VRF outdorr unit performance, while in VRF_FluidTCtrl model the *AirConditioner:VariableRefrigerantFlow* object is used.
+
```
outdorr is misspelled, s.b. outdoor

rongpengzhang replied:
Fixed.

rraustad commented on doc/src/docs/EngineeringReference/13c-EncyclopaedicRefs.md:
```diff
@@ -5362,13 +5373,1052 @@ where:
+b. Update <span>${T_e}'$</span> to meet the required evaporative capacity, using equations described in Step 2c.4a.
+
+c. Update <span>$T_e$</span> to meet the updated <span>${T_e}'$</span>. Note that due to the <span>$T_e$</span> updates, the refrigerant state and flow rate are changed and thus the piping loss analysis should also re-performed (Step 2c.1). So is the calculation of <span>$C_{cap,operation}$</span> (Step 2c.2-2c.3).
+
```
should also be repeated?

rongpengzhang replied:
Yes, the piping loss and modification factor calculations are repeated here.

rraustad replied:
I meant ... should you change the wording to say
"should also be repeated:

rongpengzhang replied:
Got it.

rraustad commented on doc/src/docs/EngineeringReference/13c-EncyclopaedicRefs.md:
```diff
@@ -5362,13 +5373,1052 @@ where:
+##### Step 2h.5: Modify evaporating temperature to further reduce outdoor unit capacity 
+
+If the calculated <span>$rps$</span> is lower than the minimum <span>$rps$</span> (e.g. 18 <span>$rps$</span>), it means that the zone heating load (indoor unit side) is so low that it leads to an evaporative capacity (outdoor unit side) which is even lower than the system evaporative capacity corresponding to the minimum compressor speed. In this situation, the evaporating temperature <span>$Te$ ...
+
```
superheating degree $SH$ ?

rongpengzhang replied:
Fixed.

rongpengzhang commented:
Hi @rraustad, I have finished the IO Ref and Eng Ref, and fixed the merge conflicts caused by the unit test.

There is 1 warning message (only on MacOS and Ubuntu platform): "src/EnergyPlus/DXCoils.hh:1284:31 unused variable ‘EnergyPlus::DXCoils::DXCoil’ [-Wunused-
variable]". But we know this variable is widely used in both DXCoils.cc and other cc files. Also, this line is not new added. Do you have any idea about this?

Hi @Myoldmopar , when I prepare the Eng Reference, I notice that the equations can be automatically labeled but the figures cannot. So I use "VRF_FluidTCtrl-1" to label the new figures I added. If you have other suggestions, I can replace these labels.

rraustad commented on doc/src/docs/EngineeringReference/13c-EncyclopaedicRefs.md:
```diff
@@ -5362,13 +5373,1052 @@ where:
+c. <span>$SH$</span> for each indoor unit can be updated using the equations shown in Step 1.2.
+
+It should be noted that, different from the corresponding step in the cooling mode (Step 2c.5), the <span>$Te$</span> and <span>$SH$</span> updates in the heating mode do not affect the refrigerant state and flow rate calculations (as shown in Step 2h.1). Therefore, the piping loss analysis does not need to be re-performed here.
+
```
does not need to be repeated here?

rongpengzhang replied:
No, no need to repeat in heating mode. This is different from the cooling mode.

rraustad commented on doc/src/docs/EngineeringReference/13c-EncyclopaedicRefs.md:
```diff
@@ -5362,13 +5373,1052 @@ where:
+
+<span>$T_{c,ref}$</span>		reference condensing temperature (°C) 
+
+<span>$N_{comp,rps}$</span>	compress power corresponding to rps (kW) 
```
compressor power...

rraustad replied:
"compress" should be "compressor"

rongpengzhang replied:
fixed.

rraustad commented on doc/src/docs/EngineeringReference/13c-EncyclopaedicRefs.md:
```diff
@@ -5362,13 +5373,1052 @@ where:
+
+<span>$Q_{in,total}$</span>	total cooling load for the zone (kW)
+
+<span>$H_{coil,in}$</span>	enthalpy of the suction air of the indoor unit (kJ/kg)
```
suction air? how about "entering air"

rongpengzhang replied:
Replaced.

rraustad commented on doc/src/docs/EngineeringReference/13c-EncyclopaedicRefs.md:
```diff
@@ -5362,13 +5373,1052 @@ where:
+
+<span>$Q_{in,total}$</span>	total cooling load for the zone (kW)
+
+<span>$T_{coil,in}$</span>	temperature of the suction air of the indoor unit (°C)
```
entering air?

rraustad replied:
change to "entering"

rongpengzhang replied:
"suction air" is replaced with "entering air"

rraustad commented on doc/src/docs/EngineeringReference/13c-EncyclopaedicRefs.md:
```diff
@@ -5362,13 +5373,1052 @@ where:
+#### Additional energy consumption by defrost and crankcase heater 
+
+There may be additional energy consumption due to the defrost operation and crankcase heater operation. These components have no impact on the heat pump operations. The calculation methods in the VRF_FluidTCtrl model are the same as those in VRF_SysCurve model. Please refer to the VRF_SysCurve section for more details.
+
```
I just changed the defrost calculation in VRF_SysCurve, did that get included?
See https://github.com/NREL/EnergyPlus/pull/5135/files

rongpengzhang replied:
I will update it.

rongpengzhang replied:
Rich, do you think it would be better to create a separated method for such calculations? Also for crankcase heater calculations.

rraustad replied:
Yes, we could move all these similar routines to a common subroutine. I 
won't make you do that here.

On 8/27/2015 2:55 PM, Rongpeng Zhang wrote:

> In doc/src/docs/EngineeringReference/13c-EncyclopaedicRefs.md 
> https://github.com/NREL/EnergyPlus/pull/5026#discussion_r38132816:
> 
> > +Set <span>$G_a$</span> at its minimum value (<span>$0.7\times$</span><span>$G_{a,rate}$</span>).
> > +
> > +The required coil surface temperature <span>$T_{fs}$</span> can be determined as described in Step 2.
> > +
> > +Given <span>$T_{fs}$</span> and <span>$T_e$</span>, <span>$SC$</span> can be determined using the equation shown in Step 1.3.
> > +
> > +\* If <span>$SC$</span> > 20°C, Set <span>$SC$</span> as 20°C. On/Off control strategy may be implemented when needed.
> > +
> > +##### Step 3h.3: Total power consumption of the indoor unit
> > +
> > +Calculate electric power consumption by the indoor unit using the existing VAV fan model in EnergyPlus. Please refer to the current EnergyPlus Engineering Reference for more details.
> > +
> > +#### Additional energy consumption by defrost and crankcase heater
> > +
> > +There may be additional energy consumption due to the defrost operation and crankcase heater operation. These components have no impact on the heat pump operations. The calculation methods in the VRF_FluidTCtrl model are the same as those in VRF_SysCurve model. Please refer to the VRF_SysCurve section for more details.
> > +
> 
> Rich, do you think it would be better to create a separated method for 
> such calculations? Also for crankcase heater calculations.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/NREL/EnergyPlus/pull/5026/files#r38132816.

## 

Richard Raustad
Florida Solar Energy Center
1679 Clearlake Road
Cocoa, FL 32922
Ph: (321)638-1454
http://www.fsec.ucf.edu/en/

Program Director
Electric Vehicle Transportation Center
http://evtc.fsec.ucf.edu/

rraustad commented on doc/src/docs/InputOutputReference/01c-InputOutputReference.md:
```diff
@@ -13648,9 +13662,16 @@ Figure 125. Variable Refrigerant Flow Schematic
+
+![VRFObjects](media/image311.png)
+*(b) VRF model based on physics (applicable for fluid temperature control)*
 
```
Why do these use the same figure? image311.png

rongpengzhang replied:
Sorry. It should be image312 for (b)

Myoldmopar commented:
@rongpengzhang 

> There is 1 warning message (only on MacOS and Ubuntu platform): "src/EnergyPlus/DXCoils.hh:1284:31 unused variable ‘EnergyPlus::DXCoils::DXCoil’ [-Wunused-
> variable]". But we know this variable is widely used in both DXCoils.cc and other cc files. Also, this line is not new added. Do you have any idea about this?

It's not saying the variable is _never_ used, it is just saying it's not used in one of the routines.  It is currently being brought into the SimulateVRF function, but not used, [here](https://github.com/NREL/EnergyPlus/blob/91977808_LBNL_NewVRFModel/src/EnergyPlus/HVACVariableRefrigerantFlow.cc#L241).  The error message isn't super clear, but well, that's C++...  Take out that line and you'll see it compile without any warnings.

rraustad commented on doc/src/docs/InputOutputReference/01c-InputOutputReference.md:
```diff
@@ -14086,6 +14107,375 @@ AirConditioner:VariableRefrigerantFlow,
+#### Field: Zone Terminal Unit List Name
+
+This alpha field defines the name of the zone terminal unit list. The name specified here should match the name of a valid ZoneTerminalUnitList object. In addition, each name specified in this list should match the name of a valid ZoneHVAC:TerminalUnit: VariableRefrigerantFlow object. All terminal units connected to this heat pump must be listed in this ZoneTerminalUnitList object.
+
```
ZoneHVAC:TerminalUnit: VariableRefrigerantFlow

1) space in name, and it should be
2) ZoneHVAC:TerminalUnit:VariableRefrigerantFlow:FluidTemperatureControl

rongpengzhang replied:
As what we discussed in the NFP review meeting, we use the existing object ZoneHVAC:TerminalUnit:VariableRefrigerantFlow in VRF-FluidTCtrl model. The object connections is shown in image312.

rraustad replied:
My mistake, name is correct.

rraustad commented on doc/src/docs/InputOutputReference/01c-InputOutputReference.md:
```diff
@@ -14086,6 +14107,375 @@ AirConditioner:VariableRefrigerantFlow,
+#### Field: Outdoor Unit Fan Object Type
+
+This alpha field defines the type of fan in the VRF outdoor unit. The only valid choices are *Fan:OnOff* and *Fan:ConstantVolume*. The input requirements for these fan objects are described elsewhere in this document.
+
```
Wow, you actually tied in a fan to the outdoor unit. I'll have to look and see how this was done.

rongpengzhang replied:
In fact, we only care about the fan power here.

rraustad replied:
So if a fan is not actually used, then why ask for the fan name? Unless 
a fan is actually connected in the system you should ask for a type and 
name.

On 8/27/2015 3:14 PM, Rongpeng Zhang wrote:

> In doc/src/docs/InputOutputReference/01c-InputOutputReference.md 
> https://github.com/NREL/EnergyPlus/pull/5026#discussion_r38134919:
> 
> > +#### Field: Variable Evaporating Temperature Maximum for Indoor Unit
> > +
> > +This numeric field defines the maximum evaporating temperature for the indoor unit when VRF runs at cooling mode. This field is required if Refrigerant Temperature Control Algorithm is VariableTemp. If this field is blank, the default value of 13.0ºC is used.
> > +
> > +#### Field: Variable Condensing Temperature Minimum for Indoor Unit
> > +
> > +This numeric field defines the minimum condensing temperature for the indoor unit when VRF runs at heating mode. This field is required if Refrigerant Temperature Control Algorithm is VariableTemp. If this field is blank, the default value of 42.0ºC is used.
> > +
> > +#### Field: Variable Condensing Temperature Maximum for Indoor Unit
> > +
> > +This numeric field defines the maximum condensing temperature for the indoor unit when VRF runs at heating mode. This field is required if Refrigerant Temperature Control Algorithm is VariableTemp. If this field is blank, the default value of 46.0ºC is used.
> > +
> > +#### Field: Outdoor Unit Fan Object Type
> > +
> > +This alpha field defines the type of fan in the VRF outdoor unit. The only valid choices are _Fan:OnOff_ and _Fan:ConstantVolume_. The input requirements for these fan objects are described elsewhere in this document.
> > +
> 
> In fact, we only care about the fan power here.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/NREL/EnergyPlus/pull/5026/files#r38134919.

## 

Richard Raustad
Florida Solar Energy Center
1679 Clearlake Road
Cocoa, FL 32922
Ph: (321)638-1454
http://www.fsec.ucf.edu/en/

Program Director
Electric Vehicle Transportation Center
http://evtc.fsec.ucf.edu/

rongpengzhang replied:
At the beginning, all the VRF OU fan information was put in the VRF object. These values are used to calculate the fan power.

But then we separated the fan information and put them in a fan object. First, this is more organized for the user inputs. Second, it will be easier to maintain in the future, if the system adds more controls on the OU fan, or the fan is improved to variable speed. Do you think it is fine?

rraustad replied:
"But then we separated the fan information and put them in a fan object"  
Was this discussed?

@mjwitte , we shouldn't have a fan object in the input only for fan details.

mjwitte replied:
@rongpengzhang Is the fan object simulation being called from the VRF outdoor unit?  

Where is the VRF outdoor unit getting it's "outdoor air temperature" from?  (weather file ODB, or the fan object inlet node?)

It is outside the normal realm of EnergyPlus input to have a fan with dangling inlet/outlet nodes that lead nowhere, but I understand the motivation.

How will the fan flow rate autosize?  It must know that it's in a VRF outdoor air unit and then there will need to be some inputs to allow the user to specify how it's sized.

rongpengzhang replied:
@mjwitte @rraustad 

Currently the fan object is only used to store the fan information. Such information is used to calculate the fan power. The VRF-FluidTCtrl model does not call the Fan object simulation. The fan flow rate cannot be autosized due to the dangling inlet/outlet nodes.

Maybe I can add some more descriptions in the VRF IO reference, stating that the fan cannot be autosized? Or, do you suggest to move the fan information back to VRF object?

mjwitte replied:
@rongpengzhang No, this is not going to be good.  Besides the fact that the separate object implies more than is really there, if the flow can't be autosized, then it's a big burden on the user to come up with something and it limits usefulness of the model for parametrics.  Does mfr data even give the airflow rate for the outdoor fan?  We may need to talk on the phone on Friday.  Let's move this discussion to an e-mail thread.

rongpengzhang replied:
@mjwitte 

I think I can move the fan information back to VRF object, as what it was originally designed. It is just one field:
    170,                     !- VRF Condenser Fan Power {W}

I think this value can be obtained from the mfr data. If you think this is fine, I will do that and get this branch merged.

I moved the fan information from VRF object to fan object because I thought it may benefit the future development/maintenance if  the system would add more control logic on the OU fan operation, or the fan is improved from constant speed to variable speed. For now, it is not really necessary to separate it into a fan object, since it is constant speed, and is not located in any air loops. Maybe we can further discuss this later after this busy season. How do you think about it?

mjwitte replied:
I think for now that would be best, but it needs to be autosizable or 
scaleable somehow.  Other DX equipment just includes this in the overall 
COP, so you could just drop it altogether for now and indicate that the 
COP includes the outdoor fan power.  It's fine to have it separate, but 
should be W per compressor like the crankcase heat, or W per capacity 
unit.  I don't even see a total capacity for the outdoor unit?

On 8/27/2015 4:34 PM, Rongpeng Zhang wrote:

> In doc/src/docs/InputOutputReference/01c-InputOutputReference.md 
> https://github.com/NREL/EnergyPlus/pull/5026#discussion_r38148968:
> 
> > +#### Field: Variable Evaporating Temperature Maximum for Indoor Unit
> > +
> > +This numeric field defines the maximum evaporating temperature for the indoor unit when VRF runs at cooling mode. This field is required if Refrigerant Temperature Control Algorithm is VariableTemp. If this field is blank, the default value of 13.0ºC is used.
> > +
> > +#### Field: Variable Condensing Temperature Minimum for Indoor Unit
> > +
> > +This numeric field defines the minimum condensing temperature for the indoor unit when VRF runs at heating mode. This field is required if Refrigerant Temperature Control Algorithm is VariableTemp. If this field is blank, the default value of 42.0ºC is used.
> > +
> > +#### Field: Variable Condensing Temperature Maximum for Indoor Unit
> > +
> > +This numeric field defines the maximum condensing temperature for the indoor unit when VRF runs at heating mode. This field is required if Refrigerant Temperature Control Algorithm is VariableTemp. If this field is blank, the default value of 46.0ºC is used.
> > +
> > +#### Field: Outdoor Unit Fan Object Type
> > +
> > +This alpha field defines the type of fan in the VRF outdoor unit. The only valid choices are _Fan:OnOff_ and _Fan:ConstantVolume_. The input requirements for these fan objects are described elsewhere in this document.
> > +
> 
> @mjwitte https://github.com/mjwitte
> 
> I think I can move the fan information back to VRF object, as what it 
> was originally designed. It is just one field:
> 170, !- Rated Condenser Fan Power {W}
> 
> I think this value can be obtained from the mfr data. If you think 
> this is fine, I will do that and get this branch merged.
> 
> I moved the fan information from VRF object to fan object because I 
> thought it may benefit the future development/maintenance if the 
> system would add more control logic on the OU fan operation, or the 
> fan is improved from constant speed to variable speed. For now, it is 
> not really necessary to separate it into a fan object, since it is 
> constant speed, and is not located in any air loops. Maybe we can have 
> further discuss this later after this busy season. How do you think 
> about it?
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/NREL/EnergyPlus/pull/5026/files#r38148968.

rongpengzhang replied:
@mjwitte  The system capacity is now calculated by the compressor curves (Loading Index i Evaporative Capacity Function of Temperature Curve). The output of the curve is absolute capacity (W), not a multiplier.

Tianzhen and I tried to design a field for the rated capacity, and let the curve to calculate a multiplier that can be applied to the rated capacity. However, our manufacturer partner states that the curves are only applicable for a specific VRF system model, not a series of models. Therefore, we don't have a rated capacity field in the VRF object and thus the system are not autosizable. Users need to do the sizing, select a specific system model, provide the model curves, and then perform the simulation. I think we discussed about this in the NFP review meeting.

We will develop the VRF-HR model in the following months. Tianzhen and I will double check with the manufacturers to see whether it is possible to set up a rated capacity. If so, we will push to set up a rated capacity field for that and activate the autosize function for both the HP and HR model.

mjwitte replied:
The curves aren't normalized?  <major-big-sigh>  That's not like 
anything else in EnergyPlus (except the refrigeration).  So, if someone 
wants to use this model, they have to simulate once to find out the 
design load, then scale all the curves to provide enough capacity?  How 
is that useable?
@hongtz68
With any kind of equipment the curves vary for different size units, but 
we've always strived to make the equipment scalable.

hongtz68 replied:
@mjwitte 

We indeed debate this a dozen times with Daikin engineers. I was pushing to have consistent models like existing VRF in EnergyPlus that enables autosizing. What Daikin engineers argued is the new model is component based and their VRF systems are custom made for combination of outdoor and indoor units, so they do not have normalized curves to represent a family of products. We looked at some Danfoss compressors (used in Daikin outdoor units) and indeed the curves vary with size/capacity. This is the basis of the current implementation.

We realize this sacrifices some usability and are developing new methods to improve the model to enable autosizing. This is part of our current work with Daikin on the VRF heat recovery systems. We hope to get the current implementation of VRF heat pump system (actually implemented in a custom Fortran version 7.2) in 8.4 release as there are users waiting for it. Then improving the autosizing feature as part of the coming VRF heat recovery implementation, hopefully for EnergyPlus 9.0 release.

rongpengzhang commented:
@Myoldmopar Thanks for pointing it out.

How did you identify that it is the SimulateVRF method that cause the warning? Check manually, or use some other secret weapon? Just curious ...

rraustad commented on doc/src/docs/InputOutputReference/01c-InputOutputReference.md:
```diff
@@ -14086,6 +14107,375 @@ AirConditioner:VariableRefrigerantFlow,
+
+This alpha field defines the name of a BiQuadratic curve for the VRF operational mode corresponding to the i-th loading index. It parameterizes the variation of compressor power as a function of operating conditions, i.e., evaporating and condensing temperatures. The output of this curve is the compressor power.
+Following is an example input for a AirConditioner:VariableRefrigerantFlow system.
+
```
AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl

rongpengzhang replied:
Fixed. Thanks!

rraustad replied:
My mistake, name is correct.

rraustad commented on doc/src/docs/InputOutputReference/01c-InputOutputReference.md:
```diff
@@ -14086,6 +14107,375 @@ AirConditioner:VariableRefrigerantFlow,
+#### Field: Compressor maximum delta Pressure
+
+This numeric field defines the maximum pressure increase that the compressor can provide. The value should be greater than 0. If this field is blank, the default value of 4500 kPa is used.
+
```
Units are Pa. Change this to read: 4,500,000 Pa so users won't get confused.

rongpengzhang replied:
Fixed.

Myoldmopar commented:
> How did you identify that it is the SimulateVRF method that cause the warning? Check manually, or use some other secret weapon? Just curious ...

It has to do with [midichlorians](https://en.wikipedia.org/wiki/The_Force_%28Star_Wars%29).  At some point, you just become able to communicate with the compiler on a higher, more zen-like, or even force-like, level.  @amirroth could speak to this.  Anytime CI produces a warning, the force awakens and starts informing of the problem and how to solve it.

.
.
.
.
.
.
.
.
**or**
.
.
Maybe I tried to compile it myself and it gave me a clue

```
[ 47%] Building CXX object src/EnergyPlus/CMakeFiles/energypluslib.dir/HVACVariableRefrigerantFlow.cc.o
In file included from /.../src/EnergyPlus/HVACVariableRefrigerantFlow.cc:27:0:
/.../src/EnergyPlus/DXCoils.hh: In function ‘void EnergyPlus::HVACVariableRefrigerantFlow::SimulateVRF(const string&, int, bool, Real64&, Real64&, int&)’:
/.../src/EnergyPlus/DXCoils.hh:1284:31: warning: unused variable ‘EnergyPlus::DXCoils::DXCoil’ [-Wunused-variable]
  extern Array1D< DXCoilData > DXCoil;
```

rraustad commented on doc/src/docs/InputOutputReference/01c-InputOutputReference.md:
```diff
@@ -14192,6 +14582,25 @@ Alternate Fuel types (e.g., FuelType = NaturalGas):
+* VRF Heat Pump Indoor Unit Condensing Temperature at Heating Mode [C]
+
+* VRF Heat Pump Outdoor Unit Evaporating Temperature at Heating Mode [C]
+
```
If you know the evaporating temperature, why are you using the same defrost model? I know the answer, there are no models. Just pointing out the question.

rongpengzhang replied:
We will take a deeper look at this when we do the VRF-HR model in the following months. Maybe we can develop a new defrost model then. Thanks for the suggestion!

rongpengzhang commented:
Nice idea @Myoldmopar! I will enforce my Zen power ...

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -4278,6 +4278,239 @@ This is the total heating output of the DX coil in Joules over the time step bei
+    autosize,                !- Rated Total Cooling Capacity {W}
+    autosize,                !- Rated Sensible Heat Ratio
+	3,                       !- Indoor Unit Reference Superheating Degrees {C}    
+	IUEvapTempCurve,         !- Indoor Unit Evaporating Temperature Function of Superheating Curve Name    
```
picky but aline these 2 lines with others

rongpengzhang replied:
Tab character again... I will fix that.

Just to confirm, the new EPlus custom is to use space instead of tab, right?

rraustad replied:
Not sure in Markdown. I think you can use tabs as long as it looks good.

On 8/27/2015 3:25 PM, Rongpeng Zhang wrote:

> In doc/src/docs/InputOutputReference/01d-InputOutputReference.md 
> https://github.com/NREL/EnergyPlus/pull/5026#discussion_r38135973:
> 
> > +#### Field: Name of Water Storage Tank for Condensate Collection
> > +
> > +This field is optional. It is used to describe where condensate from the coil is collected. If blank or omitted, then any coil condensate is discarded. Enter the name of Water Storage Tank object defined elsewhere and the condensate will then be collected in that tank.
> > +
> > +Following is an example input for a Coil:Cooling:DX:VariableRefrigerantFlow:FluidTemperatureControl object.
> > +
> > +```idf
> > -  Coil:Cooling:DX:VariableRefrigerantFlow:FluidTemperatureControl,
> > -    TU1 VRF DX Cooling Coil, !- Name
> > -    VRFAvailSched,           !- Availability Schedule Name
> > -    TU1 VRF DX CCoil Inlet Node,  !- Coil Air Inlet Node
> > -    TU1 VRF DX CCoil Outlet Node,  !- Coil Air Outlet Node
> > -    autosize,                !- Rated Total Cooling Capacity {W}
> > -    autosize,                !- Rated Sensible Heat Ratio
> > -   3,                       !- Indoor Unit Reference Superheating Degrees {C}
> > -   IUEvapTempCurve,         !- Indoor Unit Evaporating Temperature Function of Superheating Curve Name
> 
> Tab character again... I will fix that.
> 
> Just to confirm, the new EPlus custom is to use space instead of tab, 
> right?
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/NREL/EnergyPlus/pull/5026/files#r38135973.

## 

Richard Raustad
Florida Solar Energy Center
1679 Clearlake Road
Cocoa, FL 32922
Ph: (321)638-1454
http://www.fsec.ucf.edu/en/

Program Director
Electric Vehicle Transportation Center
http://evtc.fsec.ucf.edu/

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -4278,6 +4278,239 @@ This is the total heating output of the DX coil in Joules over the time step bei
+
+* HVAC,Average, Cooling Coil VRF Evaporating Temperature [C]
+
+*  HVAC,Average, Cooling Coil VRF Super Heating Degrees [C]
```
extra space at beginning of line

rongpengzhang replied:
Fixed. Thanks!

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -4278,6 +4278,239 @@ This is the total heating output of the DX coil in Joules over the time step bei
+#### Field: Name
+
+This alpha field defines a unique user-assigned name for an instance of a VRF DX heating coil. Any reference to this DX heating coil by another object will use this name. This heating coil name must be entered in the AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl object. No other system type uses this specific coil(Ref. Coil:Heating:DX:VariableRefrigerantFlow:FluidTemperatureContr ...
+
```
You don't need the (Ref...) here

rraustad replied:
Fix this one also.

rongpengzhang replied:
Fixed.

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -4278,6 +4278,239 @@ This is the total heating output of the DX coil in Joules over the time step bei
+#### Field: Name
+
+This alpha field defines a unique user-assigned name for an instance of a VRF DX cooling coil. Any reference to this DX cooling coil by another object will use this name. This cooling coil name must be entered in the AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl object. No other system type uses this specific coil (Ref. Coil:Cooling:DX:VariableRefrigerantFlow:FluidTemperatureCont ...
+
```
You don't need the (Ref...) here

rongpengzhang replied:
Fixed.

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -4278,6 +4278,239 @@ This is the total heating output of the DX coil in Joules over the time step bei
+#### Field: Availability Schedule
+
+This alpha field defines the name of the schedule (ref: Schedule) that denotes whether the DX heating coil can run during a given time period. A schedule value greater than 0 (usually 1 is used) indicates that the unit can be on during the time period. A value less than or equal to 0 (usually 0 is used) denotes that the unit must be off for the time period. If this field is blank, the schedule ha ...
+
```
If the field is blank, there is no schedule. Change to:

If this field is blank the unit is always available.

rraustad replied:
Revise text here.

rongpengzhang replied:
Revised.

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -4278,6 +4278,239 @@ This is the total heating output of the DX coil in Joules over the time step bei
+    TU1 VRF DX HCoil Outlet Node,  !- Coil Air Outlet Node
+    autosize,                !- Rated Total Heating Capacity {W}
+    5,                       !- Indoor Unit Reference Subcooling Degrees {C}    
+	IUCondTempCurve;         !- Indoor Unit Condensing Temperature Function of Subcooling Curve Name
```
align text

rongpengzhang replied:
Fixed.

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -4278,6 +4278,239 @@ This is the total heating output of the DX coil in Joules over the time step bei
+This is the subcooling degrees of the VRF system operating at heating mode. This value is manipulated by each VRF terminal unit to adjust the heating capacity of the coil considering the load conditions of the zone. It affects the heating coil surface temperature and thus the heating capacity of the coil.
+
+
+
```
Whew. Finished reviewing docs. Yeah!

rongpengzhang replied:
Thanks, Rich! Really appreciate it!

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52924,6 +53042,369 @@ AirConditioner:VariableRefrigerantFlow,
        \note Enter the time constant used to model the transition
        \note from cooling only mode to heat recovery mode
 
```
How did this get here? Git must be confused.

Myoldmopar replied:
Not sure what you mean.  The two lines above, or the new idd definition below?

rraustad commented:
@Myoldmopar I have gone through this entirely.
does NREL have to commit new features (for contractual reasons) or can I?

Myoldmopar commented:
As reviewer, you are free to merge once he addresses all of your comments and you are satisfied, and @mjwitte has reviewed the IDD changes :+1:

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -52924,6 +53042,369 @@ AirConditioner:VariableRefrigerantFlow,
+       \default 2
+       \note Enter the total number of compressor. This input is used only for crankcase
+       \note heater calculations.
+  N23, \field Ratio of Compressor Size to Total Compressor Capacity
```
@rongpengzhang This field seems redundant. Isn't it just the inverse of "Number of Compressor"?

rongpengzhang replied:
@mjwitte 
The crankcase heater calculation in the new VRF  model is the same as that in the existing VRF model. It seems these two fields are different:

if ( VRF( VRFCond ).**NumCompressors** > 1 ) {
                UpperStageCompressorRatio = ( 1.0 - VRF( VRFCond ).**CompressorSizeRatio** ) / ( VRF( VRFCond ).NumCompressors - 1 );
                for ( Stage = 1; Stage <= VRF( VRFCond ).NumCompressors - 2; ++Stage ) {
                    if ( VRF( VRFCond ).VRFCondPLR < ( VRF( VRFCond ).CompressorSizeRatio + Stage \* UpperStageCompressorRatio ) ) {
                        VRF( VRFCond ).CrankCaseHeaterPower += VRF( VRFCond ).CCHeaterPower;
                    }
                }
            }

rongpengzhang committed:
VRF-HP: address comments on documentation

rongpengzhang committed:
VRF-HP: outdoor unit fan

Move the OU fan information from fan object to the VRF object to address
Mike's comments. Reason is that the fan is not in any air loop, and fan
simulation is not called. Fan object was designed only to store fan
related information.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -52924,6 +53042,368 @@ AirConditioner:VariableRefrigerantFlow,
+       \type real
+       \note Minimum compressor speed 
+       \note unit in Rev/Sec	
+       \minimum> 0   
```
Units should be declared as \units rev/sec (similar to the existing rev/min).  Also add to the units list at the top of the idd.  Same for all speed fields.

rongpengzhang replied:
Got it.

rongpengzhang commented:
@mjwitte The OU fan flow rate is used to calculate the air temperature leaving outdoor unit heat exchanger, considering the heat released/absorbed by the heat exchanger. The air temperature is further used in determining Tc (cooling mode) and Te (heating mode). The value of 3 m3/s was from the manufacturer data. I will double check with that.

rongpengzhang commented:
Hi @mjwitte, just got confirmed: 3 m3/s lies in the normal OU fan airflow range (6000-7000 cfm)

rongpengzhang committed:
VRF-HP: Normalize compressor curves

rongpengzhang committed:
VRF-HP: documentation errors

rongpengzhang committed:
IDD format

rongpengzhang committed:
Minor changes

rongpengzhang committed:
Merge branch 'develop' into 91977808_LBNL_NewVRFModel

rongpengzhang commented:
@rraustad @mjwitte I have updated the codes to address your recent comments. Now the compressor capacity and power curves are normalized. The documentation and example files are updated correspondingly. Thanks!

rongpengzhang commented:
@rraustad Tests failed on the two Windows-7 platforms, but it seems that the problem is not from the codes. I checked other branches and they met the same problems. Do you have any ideas?

rongpengzhang commented:
@rraustad It just showed that All checks have passed. If you have no further comments, please get this branch merged. Thanks!

mjwitte commented:
@rongpengzhang The windows CI machines are down - should get fixed tomorrow.  We will make another final review pass and test, so it may be a few days before it's merged.

Myoldmopar commented:
@mjwitte @rraustad As I recall, this should be ready to merge.  There are a few merge conflicts.  I started to work them out, but _quickly_ decided it would be bad for me to try to do that.  Since there are other VRF related pull requests waiting to be merged, maybe consider a chat to plan out which order these should drop in, as there is a potential for merge conflicts when each one drops in.

mjwitte commented:
@Myoldmopar  Beside the merge conflicts, @rongpengzhang graciously made some requested changes that we have not had a change to re-review yet.

rongpengzhang commented:
@Myoldmopar I can deal with the merge conflicts for this branch, after @mjwitte and @rraustad have the final check. I know you are pretty busy these days.

mjwitte commented:
@rongpengzhang 
1. New inputs look good.  
2. However, IDF Editor will need to be changed to understand rev/sec.  If this is the common usage for this application (rather than rev/min) then I will let @JasonGlazer know.  If rev/min would be acceptable it might be better to use that instead of introducing a new one.  I have no strong opinion on this.
3. Built this branch and ran the example file as-is and as an annual run.  The errror output is enormous in both cases (36k lines and 856k lines).  There are two common types of errors:

```
   ** Warning ** Coil:Heating:DX:VariableRefrigerantFlow:FluidTemperatureControl "TU1 VRF DX HEATING COIL" - Air volume flow rate per watt of rated total heating capacity is out of range at 2.672E-005 m3/s/W.
   **   ~~~   **  During Warmup, Environment=Miami Intl Ap FL USA TMY3 WMO#=722020, at Simulation time=01/01 02:15 - 02:30
   **   ~~~   ** Expected range for VolumeFlowPerRatedTotalCapacity=[2.684E-005--8.056E-005]
   **   ~~~   ** Possible causes include inconsistent air flow rates in system components or
   **   ~~~   ** inconsistent supply air fan operation modes in coil and unitary system objects.

   ** Warning ** GetSupHeatEnthalpyRefrig: Refrigerant [R410A] is saturated at the given conditions, saturated enthalpy at given temperature returned. **
   **   ~~~   ** ...Called From:VRFCondenser
   **   ~~~   ** Refrigerant temperature = -12.13
   **   ~~~   ** Refrigerant pressure = 1261139
   **   ~~~   ** Returned Enthalpy value = 417250.501
   **   ~~~   **  Environment=Miami Intl Ap FL USA TMY3 WMO#=722020, at Simulation time=01/01 02:30 - 02:32
```

The flow per capacity warnings are set up as recurring warnings, but I would not expect to see these in an example file.  There is also a sizing warning about capacity vs flow.  if these units typically operate in a wider range of cfm/ton, then the limits should be changed for this coil type.  If not, then the inputs should be fixed.

The bigger problem is the GetSupHeatEnthalpyRefrig warnings.  These shouldn't be happening in the first place, but when they do, they should only report a few times and then accumulate as a recurring warning.
1. Also, example file annual run has over 2000 hours of cooling setpoint not met.  

@Myoldmopar @rraustad If the warnings and setpoint not met issues can't be addressed today, I'm ok with this going in and resolving those under a new issue next week.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -42754,6 +42779,100 @@ Coil:Heating:DX:VariableRefrigerantFlow,
+  A5;  \field Indoor Unit Condensing Temperature Function of Subcooling Curve Name    
+       \required-field   
+       \type object-list
+       \object-list QuadraticCurves	  
```
@rongpengzhang IDF Editor doesn't like the trailing spaces on these \object-list references. I only found that on these two, but there's one somewhere else that IDF Editor is complaining about.

rongpengzhang replied:
Fixed.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -52934,6 +53053,400 @@ AirConditioner:VariableRefrigerantFlow,
+  A7 , \field Outdoor Unit Condensing Temperature Function of Subcooling Curve Name    
+       \required-field   
+       \type object-list
+       \object-list QuadraticCurves	 
```
@rongpengzhang It's these two - please delete trailing spaces and/or tabs here.

rongpengzhang replied:
Got it.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -33428,6 +33429,14 @@ ZoneHVAC:TerminalUnit:VariableRefrigerantFlow,
+        \note AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl
+        \note is used to model VRF outdoor unit
         \default Fan:ConstantVolume
   A8 ,  \field Supply Air Fan Object Name
```
@rongpengzhang The object-list needs to change now to FansCVandOnOffandVAV

rongpengzhang replied:
Fixed.

rongpengzhang commented:
@mjwitte Thanks for the review and advice!

For your 2nd comment, I think it is fine to use rev/min. It's not difficult for the user to translate their manufacturer data from rev/sec to rev/min. We don't need to bother @JasonGlazer.

For the 3rd comment, I'd prefer to resolve them under a new issue next week. I may need some time to get some manufacturer data to address the flow per capacity warnings. 

I will address all your other comments today and let you know.

Rongpeng

mjwitte commented:
@rongpengzhang OK.  New issue posted #5167

rongpengzhang committed:
VRF-HP: final updates

Delete Refrence COP from new VRF object, since Rated Evap Capacity and
Rated Power are newly added.

rongpengzhang committed:
Merge branch 'develop' into 91977808_LBNL_NewVRFModel

Conflicts addressed for:
	src/EnergyPlus/DXCoils.cc
	src/EnergyPlus/DXCoils.hh
	src/EnergyPlus/DataHVACGlobals.cc
	src/EnergyPlus/HVACVariableRefrigerantFlow.cc
	src/EnergyPlus/HVACVariableRefrigerantFlow.hh
	tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc

rongpengzhang committed:
VRF-HP: Address Mike's new comments and Merge Conflict

Delete new unit to rev/sec, using existing rev/min instead.
Merge Conflict Addressed

rongpengzhang commented:
Hi @Myoldmopar ,

Thanks for creating #5171 and the tips. I have addressed all the merge conflicts in my branch. I didn't use the #5171 branch, since it doesn't address all the files with conflicts. But I paid attention to the two points you mentioned. Now #5171 is closed and deleted as what you suggested.

I have built and run successfully at my local machine. Let's see the CI results. If everything is OK, I will notify Mike to get it merged.

Thanks,
Rongpeng

mjwitte commented:
i386-Windows CI appears to be stuck.  All other tests are OK, so merging this.

rongpengzhang commented:
Thanks for the all review and advice, @mjwitte

mjwitte commented:
@Myoldmopar Didn't notice this before merging into develop.  Triggered conflicts with my branch.  But line endings seem to be correct here.
doc/src/docs/InputOutputReference/01b-InputOutputReference.md
20,730 additions, 20,730 deletions not shown

rongpengzhang commented:
@mjwitte FYI, I didn't change 01b-InputOutputReference.md. All the IO Ref documentations related with this VRF-HP are put in 01c-InputOutputReference.md and 01d-InputOutputReference.md.

I checked all my recent commits, and find that the 01b-InputOutputReference.md modifications come from Commit 4c9311e "Merge branch 'develop' into 91977808_LBNL_NewVRFModel" yesterday. I don't know why this happened, but I think it is fine to just get rid of the 01b-InputOutputReference.md changed in this branch, i.e., to use the file in the develop branch.

Please let me know if there is anything I can do from my side.
